### PR TITLE
CmSoftwareUpdatePointComponent

### PR DESCRIPTION
**Pull request by NEllis280**
#### Pull Request (PR) description
    This PR adds a resource for the Software Update Point Component

#### Task list

- [x] Added an entry to the change log under the Unreleased section of the
      file CHANGELOG.md. Entry should say what was changed and how that
      affects users (if applicable), and reference the issue being resolved
      (if applicable).
- [x] Resource documentation added/updated in README.md.
- [x] Resource parameter descriptions added/updated in README.md, schema.mof
      and comment-based help.
- [x] Comment-based help added/updated.
- [x] Localization strings added/updated in all localization files as appropriate.
- [x] Examples appropriately added/updated.
- [x] Unit tests added/updated. See [DSC Community Testing Guidelines](https://dsccommunity.org/guidelines/testing-guidelines).
- [ ] Integration tests added/updated (where possible). See [DSC Community Testing Guidelines](https://dsccommunity.org/guidelines/testing-guidelines).
- [x] New/changed code adheres to [DSC Community Style Guidelines](https://dsccommunity.org/styleguidelines).

---
This change is [Reviewable](https://reviewable.io/reviews/dsccommunity/configmgrcbdsc/98)

### CODE REVIEW

File: source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1 [hunk 1]
```diff
@@ -0,0 +1,1518 @@
...
+            'MaxInstallTime-Windows'     { $uInstallTime = ($wsusSyncProp.Value)/60 }
+        }
+    }
+
+    if (-not [string]::IsNullOrEmpty($syncSchedule))
+    {
+        $schedule = Get-CMSchedule -ScheduleString $syncSchedule
+    }
+
+    $syncFailureAlert = get-cmalert | Where-Object -FilterScript {$_.Name -match "Synchronization failure alert for software update point:"}
```
Comment (jeffotterpohl) [inline source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1:141]: Get-CMAlert

File: source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1 [hunk 2]
```diff
@@ -0,0 +1,1518 @@
...
+    if ($syncFailureAlert)
+    {
+        $enableFailureAlert = $true
+    }
+    else
+    {
+        $enableFailureAlert = $false
+    }
+
+    $available = (Get-CMSoftwareUpdateCategory -fast)
```
Comment (jeffotterpohl) [inline source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1:152]: -Fast

File: source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1 [hunk 3]
```diff
@@ -0,0 +1,1518 @@
...
+                    }
+                }
+
+                if ($langSumCompare.Remove)
+                {
+                    $buildingParams += @{
+                        RemoveLanguageSummaryDetail = $langSumCompare.Remove
+                    }
+                }
+
```
Comment (jeffotterpohl) [inline source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1:664]: extra space

File: source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1 [hunk 4]
```diff
@@ -0,0 +1,1518 @@
...
+                if ($langFileCompare.Remove)
+                {
+                    $buildingParams += @{
+                        RemoveLanguageUpdateFile = $langFileCompare.Remove
+                    }
+                }
+
```
Comment (jeffotterpohl) [inline source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1:702]: extra space

File: source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1 [hunk 5]
```diff
@@ -0,0 +1,1518 @@
...
+        [ValidateRange(1,99)]
+        [UInt32]
+        $WaitMonthForFeature,
+
+        [Parameter()]
+        [String]
+        $Start,
+
+        [Parameter()]
+        [ValidateSet('MonthlyByDay','MonthlyByWeek','Weekly','Days','None')]
```
Comment (jeffotterpohl) [inline source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1:1098]: Remove None, from here and all validate sets including readme, schema etc.

File: source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1 [hunk 6]
```diff
@@ -0,0 +1,1518 @@
...
+        }
+
+        $mainState = Test-DscParameterState @testParams -TurnOffTypeChecking -Verbose
+
+        if (-not [string]::IsNullOrEmpty($mainState) -and $mainState -eq $false)
+        {
+            $result = $false
+        }
+    }
+    else {
```
Comment (jeffotterpohl) [inline source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1:1243]: new line {

File: source/Modules/ConfigMgrCBDsc.ReverseDsc/ConfigMgrCBDsc.ReverseDsc.psm1 [hunk 7]
```diff
@@ -3744,7 +6246,8 @@ function Set-ConfigMgrCBDscReverse
             'MaintenanceWindow','ManagementPoint','NetworkDiscovery','PullDistributionPoint',
             'PxeDistributionPoint','ReportingServicesPoint','SecurityScopes','ServiceConnection',
             'SiteMaintenance','SiteSystemServer','SoftwareDistributionComponent','SoftwareupdatePoint',
-            'StatusReportingComponent','SystemDiscovery','UserDiscovery','ConfigFileOnly')]
+            'SoftwareupdatePointComponent','StatusReportingComponent','SystemDiscovery','UserDiscovery',
```
Comment (jeffotterpohl) [inline source/Modules/ConfigMgrCBDsc.ReverseDsc/ConfigMgrCBDsc.ReverseDsc.psm1:6249]: SoftwareUpdatePointComponent

File: source/Modules/ConfigMgrCBDsc.ReverseDsc/ConfigMgrCBDsc.ReverseDsc.psm1 [hunk 8]
```diff
@@ -3756,7 +6259,7 @@ function Set-ConfigMgrCBDscReverse
             'MaintenanceWindow','ManagementPoint','NetworkDiscovery','PullDistributionPoint',
             'PxeDistributionPoint','ReportingServicesPoint','SecurityScopes','ServiceConnection',
             'SiteMaintenance','SiteSystemServer','SoftwareDistributionComponent','SoftwareupdatePoint',
-            'StatusReportingComponent','SystemDiscovery','UserDiscovery')]
+            'SoftwareupdatePointComponent','StatusReportingComponent','SystemDiscovery','UserDiscovery')]
```
Comment (jeffotterpohl) [inline source/Modules/ConfigMgrCBDsc.ReverseDsc/ConfigMgrCBDsc.ReverseDsc.psm1:6262]: SoftwareUpdatePointComponent

File: tests/Unit/ConfigMgrCBDsc.ReverseDsc.tests.ps1 [hunk 9]
```diff
@@ -1849,6 +1849,224 @@ InModuleScope $script:subModuleName {
...
+                            Name         = 'Start'
+                            PropertyType = '[string]'
+                            IsMandatory  = $false
+                            Values       = '{}'
+                        }
+                        @{
+                            Name         = 'ScheduleType'
+                            PropertyType = '[string]'
+                            IsMandatory  = $false
+                            Values       = '{MonthlyByDay, MonthlyByWeek, Weekly, Days, None}'
```
Comment (jeffotterpohl) [inline tests/Unit/ConfigMgrCBDsc.ReverseDsc.tests.ps1:2018]: need to remove none

File: source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1 (hunk 1 above)
Comment (NEllis280) [reply source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1]: Changed

File: source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1 (hunk 2 above)
Comment (NEllis280) [reply source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1]: Changed

File: source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1 (hunk 4 above)
Comment (NEllis280) [reply source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1]: Changed

File: source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1 (hunk 6 above)
Comment (NEllis280) [reply source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1]: Changed

File: source/Modules/ConfigMgrCBDsc.ReverseDsc/ConfigMgrCBDsc.ReverseDsc.psm1 (hunk 8 above)
Comment (NEllis280) [reply source/Modules/ConfigMgrCBDsc.ReverseDsc/ConfigMgrCBDsc.ReverseDsc.psm1]: Changed

File: tests/Unit/ConfigMgrCBDsc.ReverseDsc.tests.ps1 (hunk 9 above)
Comment (NEllis280) [reply tests/Unit/ConfigMgrCBDsc.ReverseDsc.tests.ps1]: Changed

File: source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.schema.mof [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    [Write, Description("Specifies the day number for MonthlyByDay schedules.")] UInt32 DayOfMonth;
+    [Write, Description("Specifies whether manual management of the WSUS signing certificate is enabled.")] Boolean EnableManualCertManagement;
+    [Write, Description("Specifies whether third-party updates are enabled on the Software Update Point Component.")] Boolean EnableThirdPartyUpdates;
+    [Write, Description("Specifies the maximum runtime, in minutes, for windows feature updates.")] UInt32 FeatureUpdateMaxRuntimeMins;
+    [Write, Description("Specifies the maximum runtime, in minutes, for Office 365 updates and windows non-feature updates.")] UInt32 NonFeatureUpdateMaxRuntimeMins;
+    [Read, Description("Identifies whether the specified site is a child site or top level site.")] String ChildSite;
+    [Read, Description("Identifies software update categories and products at the specified site.")] String AvailableCats;
```
Comment (jeffotterpohl) [inline source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.schema.mof:40]: This should be an array Cats[]

File: source/DSCResources/DSC_CMSoftwareUpdatePointComponent/en-US/DSC_CMSoftwareUpdatePointComponent.strings.psd1 [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    ProductsInEx                 = ProductsToExclude and ProductsToInclude contain to same entry {0}, remove from one of the arrays.
+    UpdateClassInEx              = UpdateClassificationsToExclude and UpdateClassificationsToInclude contain to same entry {0}, remove from one of the arrays.
+    LanguageSummaryIgnore        = LanguageSummaryDetails is set, ignoring LanguageSummaryDetailsToInclude and LanguageSummaryDetailsToExclude settings.
+    LanguageSummaryMissing       = The following LanguageSummaryDetails are missing {0}.
+    LanguageSummaryRemove        = The following LanguageSummaryDetails need to be removed {0}.
+    LanguageUpdateIgnore         = LanguageUpdateFiles is set, ignoring LanguageUpdateFilesToInclude and LanguageUpdateFilesToExclude settings.
+    LanguageUpdateMissing        = The following LanguageUpdateFiles are missing {0}.
+    LanguageUpdateRemove         = The following LanguageUpdateFiles need to be removed {0}.
+    ProductsIgnore               = Products is set, ignoring ProductsToInclude and ProductsToExclude settings.
+    ProductsMissing              = The following Products are missing {0}.
```
Comment (jeffotterpohl) [inline source/DSCResources/DSC_CMSoftwareUpdatePointComponent/en-US/DSC_CMSoftwareUpdatePointComponent.strings.psd1:22]: Can we think about adding the NOTMATCH:

File: source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1 [hunk 12]
```diff
@@ -0,0 +1,1517 @@
...
+        {
+            if ($availableLangs -notcontains $lang)
+            {
+                throw ($script:localizedData.InvalidLangs -f $lang)
+            }
+        }
+
+        foreach ($cat in $cats.Where({ $_ -ne $null }))
+        {
+            if ($State.AvailableCats -notcontains $cat)
```
Comment (jeffotterpohl) [inline source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1:519]: $state

File: source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1 [hunk 13]
```diff
@@ -0,0 +1,1517 @@
...
+        if ($availableLangs -notcontains $lang)
+        {
+            Write-Warning -Message ($script:localizedData.InvalidLangs -f $lang)
+            $result = $false
+        }
+    }
+
+    foreach ($cat in $cats.Where({ $_ -ne $null }))
+    {
+        if ($State.AvailableCats -notcontains $cat)
```
Comment (jeffotterpohl) [inline source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1:1167]: $state

File: source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1 [hunk 14]
```diff
@@ -0,0 +1,1517 @@
...
+            }
+
+            if ($updateClassificationsCompare.Remove)
+            {
+                Write-Verbose -Message ($script:localizedData.UpdateClassificationsRemove -f ($updateClassificationsCompare.Remove | Out-String))
+                $result = $false
+            }
+        }
+
+        if ($ScheduleType)
```
Comment (jeffotterpohl) [inline source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1:1475]: maybe consider a check if they have additional schedule settings but don't have scheduletype a warn in test and throw in set?

File: source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1 [hunk 15]
```diff
@@ -0,0 +1,1517 @@
...
+        [ValidateRange(1,99)]
+        [UInt32]
+        $WaitMonthForFeature,
+
+        [Parameter()]
+        [String]
+        $Start,
+
+        [Parameter()]
+        [ValidateSet('MonthlyByDay','MonthlyByWeek','Weekly','Days')]
```
Comment (jeffotterpohl) [inline source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1:1096]: should add Hours everywhere required

File: source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1 [hunk 16]
```diff
@@ -0,0 +1,1517 @@
...
+        {
+            foreach ($child in $nonChildParams)
+            {
+                if ($PSBoundParameters.ContainsKey($child))
+                {
+                    Write-Warning -Message ($script:localizedData.ChildBadParams -f $child)
+                }
+            }
+
+            if ($LanguageUpdateFiles -or $LanguageUpdateFilesToInclude -or $LanguageUpdateFilesToExclude)
```
Comment (jeffotterpohl) [inline source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1:535]: Do we want to give the ability to Null out languages?

File: source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1 [hunk 17]
```diff
@@ -0,0 +1,1565 @@
...
+
+            if ($ReportingEvent -ne $state.ReportingEvent)
+            {
+                Write-Verbose -Message ($script:localizedData.SettingValue -f 'ReportingEvent', $ReportingEvent)
+                $buildingParams += @{
+                    ReportingEvent = $ReportingEvent
+                }
+            }
+        }
+        else {
```
Comment (jeffotterpohl) [inline source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1:585]: else
{

File: source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.schema.mof (hunk 10 above)
Comment (NEllis280) [reply source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.schema.mof]: fixed.

File: source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1 (hunk 12 above)
Comment (NEllis280) [reply source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1]: fixed.

File: source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1 (hunk 14 above)
Comment (NEllis280) [reply source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1:1524]: fixed.

File: source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1 (hunk 16 above)
Comment (NEllis280) [reply source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1:535]: fixed.

File: source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1 (hunk 17 above)
Comment (NEllis280) [reply source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1]: fixed.

File: source/DSCResources/DSC_CMSoftwareUpdatePointComponent/en-US/DSC_CMSoftwareUpdatePointComponent.strings.psd1 (hunk 11 above)
Comment (NEllis280) [reply source/DSCResources/DSC_CMSoftwareUpdatePointComponent/en-US/DSC_CMSoftwareUpdatePointComponent.strings.psd1]: fixed.

File: source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1 (hunk 13 above)
Comment (NEllis280) [reply source/DSCResources/DSC_CMSoftwareUpdatePointComponent/DSC_CMSoftwareUpdatePointComponent.psm1]: fixed.

**Review by jeffotterpohl — APPROVED** [review]
Looks good to me.  Thanks for all of the updates!

**Pull request merged**